camel/models/cohere_model.py

```diff
@@ -0,0 +1,216 @@
+# =========== Copyright 2023 @ CAMEL-AI.org. All Rights Reserved. ===========
+# Licensed under the Apache License, Version 2.0 (the “License”);
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an “AS IS” BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =========== Copyright 2023 @ CAMEL-AI.org. All Rights Reserved. ===========
+import logging
+import os
+import uuid
+from typing import TYPE_CHECKING, Any, Dict, List, Optional
+
+if TYPE_CHECKING:
+    from cohere.types import NonStreamedChatResponse
+
+from camel.configs import COHERE_API_PARAMS
+from camel.messages import OpenAIMessage
+from camel.models import BaseModelBackend
+from camel.types import ChatCompletion, ModelType
+from camel.utils import (
+    BaseTokenCounter,
+    OpenAITokenCounter,
+    api_keys_required,
+)
+
+try:
+    import os
+
+    if os.getenv("AGENTOPS_API_KEY") is not None:
+        from agentops import LLMEvent, record
+    else:
+        raise ImportError
+except (ImportError, AttributeError):
+    LLMEvent = None
+
+
+class CohereModel(BaseModelBackend):
+    r"""Cohere API in a unified BaseModelBackend interface."""
+
+    def __init__(
+        self,
+        model_type: ModelType,
+        model_config_dict: Dict[str, Any],
+        api_key: Optional[str] = None,
+        url: Optional[str] = None,
+        token_counter: Optional[BaseTokenCounter] = None,
+    ):
+        import cohere
+
+        super().__init__(
+            model_type, model_config_dict, api_key, token_counter=token_counter
+        )
+        self._api_key = api_key or os.environ.get("COHERE_API_KEY")
+        self._url = url or os.environ.get("COHERE_SERVER_URL")
+
+        self._client = cohere.Client(api_key=self._api_key)
+        self._token_counter: Optional[BaseTokenCounter] = None
+
+    def _to_openai_response(
+        self, response: 'NonStreamedChatResponse'
+    ) -> ChatCompletion:
```

Comment:
cohere model also support tool calling, we need to add tool_calls data to `ChatCompletion `